private/signer/v4/v4.go

```diff
@@ -213,31 +273,40 @@ func (v4 *signer) buildCredentialString() {
 	}
 }
 
-func (v4 *signer) buildQuery() {
-	for k, h := range v4.Request.Header {
+func buildQuery(notHoist bool, allowed map[string]bool, header http.Header) url.Values {
+	query := url.Values{}
+	for k, h := range header {
 		if strings.HasPrefix(http.CanonicalHeaderKey(k), "X-Amz-") {
```

Comment:
For `notHoist` we can skip buildQuery all together. can be done outside in build.